akd/src/utils.rs

```diff
@@ -9,36 +9,42 @@
 // 2. For each node in current_nodes set, check if each child is in prefix hashmap
 // 3. If so, add child label to batch set
 
-use crate::{NodeLabel, EMPTY_LABEL, EMPTY_VALUE};
-use std::collections::HashSet;
+use crate::{Node, EMPTY_LABEL, EMPTY_VALUE};
 
-// Builds a set of all prefixes of the input labels
-pub(crate) fn build_prefixes_set(labels: &[NodeLabel]) -> HashSet<NodeLabel> {
-    let mut prefixes_set = HashSet::new();
-    for label in labels {
-        for len in 0..(label.get_len() + 1) {
-            prefixes_set.insert(label.get_prefix(len));
-        }
+/// A node with node.label.label_len = 256. Used a wrapper around a [Node] in order to
+/// help with preloading nodes during directory publish.
+#[derive(Eq)]
+pub(crate) struct MaxLengthNode(Node);
+
+impl PartialEq for MaxLengthNode {
+    fn eq(&self, other: &Self) -> bool {
+        self.0.label.label_val == other.0.label.label_val
+            && self.0.label.label_len == other.0.label.label_len
+    }
+}
+
+impl PartialOrd for MaxLengthNode {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.0.label.label_val.cmp(&other.0.label.label_val))
     }
-    prefixes_set
 }
 
-pub(crate) fn build_lookup_prefixes_set(labels: &[NodeLabel]) -> HashSet<NodeLabel> {
-    let mut lookup_prefixes_set = HashSet::new();
-    for label in labels {
-        // We need the actual node for lookup too
-        lookup_prefixes_set.insert(*label);
-        for len in 0..(label.get_len() + 1) {
-            // Sibling prefixes unfortunately do not cover all the nodes we will need for
-            // a lookup proof. Although we can figure out which nodes are exactly needed
-            // this will require basically doing a pre-lookup.
-            // Instead here we load the prefixes as well.
-            // This combination (sibling- + self-prefixes) covers all the nodes we need.
-            lookup_prefixes_set.insert(label.get_prefix(len));
-            lookup_prefixes_set.insert(label.get_sibling_prefix(len));
-        }
+impl Ord for MaxLengthNode {
```

Comment:
If we wanted to get rid of `.into()` calls, we could support the `Deref` trait on `MaxLengthNode` to be able to silently cast it back to a Node. (Not sure it's super helpful)

Reply:
For an example, `AkdLabel` and `AkdValue` support this to get the underlying bytes without much headache

Reply:
Ah, actually, turns out I can just get rid of the `MaxLengthNode` struct altogether and re-use the Ord defined on Node / NodeLabel. Thanks!